python-frontend/src/main/java/org/sonar/python/semantic/v2/types/Propagation.java

```diff
@@ -0,0 +1,100 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.semantic.v2.types;
+
+import java.util.HashSet;
+import java.util.Set;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.python.semantic.v2.SymbolV2;
+import org.sonar.python.semantic.v2.SymbolV2Utils;
+import org.sonar.python.semantic.v2.UsageV2;
+import org.sonar.python.tree.NameImpl;
+import org.sonar.python.types.v2.PythonType;
+import org.sonar.python.types.v2.UnionType;
+
+public abstract class Propagation {
+
+  final Set<SymbolV2> variableDependencies = new HashSet<>();
+  final Set<Propagation> dependents = new HashSet<>();
+
+  final SymbolV2 lhsSymbol;
+  final Name lhsName;
+
+  protected Propagation(SymbolV2 lhsSymbol, Name lhsName) {
+    this.lhsSymbol = lhsSymbol;
+    this.lhsName = lhsName;
+  }
+
+  /**
+   * This is used for AST-based type inference in try/catch statements
+   * @return true if the propagation effectively changed the inferred type of assignment LHS
+   */
+  boolean propagate(Set<SymbolV2> initializedVars) {
+    PythonType rhsType = rhsType();
+    if (initializedVars.add(lhsSymbol)) {
+      lhsSymbol.usages().stream()
+        .filter(u -> !SymbolV2Utils.isFunctionOrClassDeclaration(u))
+        .map(UsageV2::tree)
+        .filter(NameImpl.class::isInstance)
+        .map(NameImpl.class::cast)
+        .forEach(n -> n.typeV2(rhsType));
+      return true;
+    } else {
+      PythonType currentType = currentType(lhsName);
+      if (currentType == null) {
+        return false;
+      }
+      PythonType newType = UnionType.or(rhsType, currentType);
+      lhsSymbol.usages()
+        .stream()
+        .filter(u -> !SymbolV2Utils.isFunctionOrClassDeclaration(u))
+        .map(UsageV2::tree)
+        .filter(NameImpl.class::isInstance)
+        .map(NameImpl.class::cast)
+        .forEach(n -> n.typeV2(newType));
+      return !newType.equals(currentType);
+    }
+  }
+
+  boolean areDependenciesReady(Set<SymbolV2> initializedVars) {
+    return initializedVars.containsAll(variableDependencies);
+  }
+
+  Set<Propagation> dependents() {
+    return dependents;
+  }
+
+  public abstract Name lhsName();
+
+  public abstract PythonType rhsType();
+
+  static PythonType currentType(Name lhsName) {
+    return lhsName.symbolV2().usages()
+      .stream()
+      .filter(u -> !SymbolV2Utils.isFunctionOrClassDeclaration(u))
+      .map(UsageV2::tree)
```

Comment:
maybe we can simply move this part to SymbolV2Utils:
```java
public static Stream<Tree> getSymbolNonDeclarationUsageTrees(SymbolV2 symbol) {
  
  return symbol.usages()
      .stream()
      .filter(u -> !SymbolV2Utils.isFunctionOrClassDeclaration(u))
      .map(UsageV2::tree)
}
```

Reply:
Good point! I extracted this into a dedicated method.
However, I kept it in `Propagation`, as I want the filtering out of definitions to be super explicit (I think it only makes sense in this specific case). I added a comment to clarify this in the now mutualized method.